Add painless string split function (splitOnToken)

cam72cam opened the pull request:
Fixes #20952

Example Usage:
```json
{
 "pipeline": {
   "processors": [
     {
       "script": {
         "source": "ctx.ip = ctx.ip.splitOnToken(',')[0]"
       }
     },
     {
       "geoip": {
         "field": "ip",
         "target_field": "geoip"
       }
     }
   ]
 },
 "docs": [
   {
     "_source": {
       "message": "hello world",
       "ip": "130.167.201.50"
     }
   },
   {
     "_source": {
       "message": "hello world",
       "ip": "130.167.201.50, 130.167.201.51"
     }
   }
 ]
}
```

cam72cam committed:
Add painless string split function (splitOnToken)

elasticmachine commented:
Pinging @elastic/es-core-infra

jdconrad commented:
@cam72cam Thank you for submitting this.  However, in Painless we have taken great care to avoid compiling regexes in all but static cases since compilation can have very poor performance.  If you're still interested, perhaps you could re-implement this set of methods using StringTokenizer or something similar?  The method names and whitelisting both look good.

cam72cam commented:
Ah, I understand.  Even though I am escaping a regex, java is still using a dynamically compiled regex behind the scenes.  I'll take a stab at refactoring this next week.

jdconrad commented:
@cam72cam Just wanted to loop back with you and see if you still had interest in continuing this PR?  Thanks!

cam72cam commented:
I've finally got some time to get back on this.  I'll try to get the branch updated before the weekend

cam72cam committed:
Use custom string manipulation in splitOnToken

Avoid string.split due to regex compilation

cam72cam commented:
@jdconrad I've pushed my own custom implementation of this functionality.  Based on my testing it has similar performance for single char tokens and better performance on longer tokens.

jdconrad commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
+        result.add(receiver.substring(pos));
+
+        // O(N) or faster depending on implementation
+        return result.toArray(new String[0]);
```
If you make the ```new String[0]``` into ```new String[results.size()]``` the array will not have to be allocated twice.

cam72cam replied:
From what I have been reading, it's more efficient to use a empty array here:
https://stackoverflow.com/questions/5374311/convert-arrayliststring-to-string-array
https://shipilev.net/blog/2016/arrays-wisdom-ancients/

jdconrad replied:
Interesting articles.  Thank you!

jdconrad reviewed (changes requested):
@cam72cam Thanks for updating this.  It's starting to look good.  Would you please add unit tests for these two methods?

cam72cam commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
+        // Loop until we hit the limit or forever if we are passed in less than one (signifying no limit)
+        // If Integer.MIN_VALUE is passed in, it will still continue to loop down to 1 from MAX_VALUE
+        // This edge case should be fine as we are limited by receiver length (Integer.MAX_VALUE) even if we split at every char
+        for(;limit != 1; limit--) {
```
That is the desired functionality and matches String.split in terms of accepted length.  less than 1 should be treated as split as many times as you can.

jdconrad replied:
Yes, of course, brain glitch on my part.

cam72cam commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
+    }
+
+    /**
+     * Split 'receiver' by 'token' up to 'limit' times.  Any limit less than 1 is ignored.
```
See above comment.  I'd like to keep this as compatible with String.split as possible

jdconrad replied:
Same as above.

jdconrad commented:
@cam72cam Thanks for the responses.  Just need some unit tests, and then I will run our test suite against this change.

cam72cam commented:
Ok, I'll try to get some unit tests in next week

jdconrad commented:
@cam72cam Are you still interested in finishing this up with some unit tests?  Thanks!

cam72cam commented:
Yes, sorry.  Work is quite hectic and I have not had a chance to sit down for a hour or two at work and write them up.

jaymode committed:
Update TLS ciphers and protocols for JDK 11 (#41808)

* Update TLS ciphers and protocols for JDK 11 (#41385)

This commit updates the default ciphers and TLS protocols that are used
after the minimum supported JDK is JDK 11. The conditionals around
TLSv1.3 and 256-bit cipher support have been removed. JDK 11 no longer
requires an unlimited JCE policy file for 256 bit cipher support and
TLSv1.3 is supported in JDK 11+. New cipher support has been introduced
in the newer JDK versions as well. The ciphers are ordered with PFS
ciphers being most preferred, then AEAD ciphers, and finally those with
mainstream hardware support.

* Fixes for TLSv1.3 on JDK11

* fix for JDK-8212885

cam72cam committed:
Add tests for painless splitOnString

cam72cam commented:
@jdconrad I have added some tests and rebased on top of master

jdconrad commented:
@cam72cam The added tests look good.  Thank you for finishing this up.  However, I'm not sure the rebase went correctly as there are a large number of files that shouldn't have changes.  Would you please try syncing this again with master?

cam72cam committed:
painless api reference updated

cam72cam committed:
Remove some build cache files from conflict / rebase

cam72cam commented:
@jdconrad Looks like someone force pushed to origin/master a few days ago.  I rebased and cleaned up my branch.

jdconrad commented:
@cam72cam This is great, thank you!  Running tests now.

jdconrad commented:
@elasticmachine test this please

cam72cam commented:
hmm: 
```
15:29:40 Forbidden field access: java.lang.System#out [prints to System.out; should only be used for debugging, not in production code]
15:29:40   in org.elasticsearch.painless.AugmentationTests (AugmentationTests.java:235)
```

I can comment that out in the tests

jdconrad commented:
@cam72cam I missed the println when reviewing the tests here.  Please go ahead and remove that line completely, and then I will retest.

cam72cam committed:
Comment out painless string split test debug

cam72cam commented:
@jdconrad pushed

jdconrad commented:
@elasticmachine test this please

jdconrad commented:
@cam72cam Triggered another forbidden API.  I would just use the build in string concat's ```+``` here instead of format.  I don't know what your exact setup is for testing, but if you run the gradle command precommit within the modules/lang-painless directory, it should reveal any other issues related to style.

cam72cam committed:
Remove string.format from augment test

jdconrad commented:
@elasticmachine test this please

jdconrad commented:
@elasticmachine run elasticsearch-ci/1

jdconrad commented:
@cam72cam Your change has been merged.  Thank you for your contribution, and thank you for sticking with it to completion!

cam72cam commented:
Thanks for being patient with my chaotic schedule!
